WordPress/Classes/ViewRelated/Reader/ReaderSearchViewController.swift

```diff
@@ -304,16 +303,20 @@ import Gridicons
 
         let previousTopic = streamController.readerTopic
 
-        let context = ContextManager.sharedInstance().mainContext
-        let service = ReaderTopicService(managedObjectContext: context)
+        let service = ReaderTopicService(coreDataStack: ContextManager.shared)
+        service.createSearchTopic(forSearchPhrase: phrase) { topicID in
+            assert(Thread.isMainThread)
```

Comment:
Also, would passing the desired thread as parameter to `createSearchTopic(forSearchPhrase:, _:)` help?

🤔 

It wouldn't help enforce the thread being the main one, I guess, because there could always be human error, but maybe it would remove the need to that assertion?

Unless... We revisited the idea of having a (closure, `DispatchQueue`) pair object to encapsulate that "this code should run on this thread" relationship, and then specialize it to one where the `DispatchQueue` parameter is `.main`. Maybe?

Reply:
> Unless... We revisited the idea of having a (closure, DispatchQueue) pair object to encapsulate that "this code should run on this thread" relationship, and then specialize it to one where the DispatchQueue parameter is .main. Maybe?

Or maybe not! 

Forcing `createSearchTopic(forSearchPhrase:, _:)` to have a completion that runs on the main thread as a function parameter is the same as calling `.main` internally (maybe with the advantage of making it explicit?) but I think it gets the concern backwards: The service should not make assumptions on which thread it should run the given completion, I think.

Reply:
Since this function uses the `mainContext`, which is associated with the main thread, to perform a query.

> Unless... We revisited the idea of having a (closure, DispatchQueue) pair object 

Sorry, I remembered you mentioned this in another PR, I forgot to respond. Yes, I think that would be a good idea. I can definitely start trying it out. I think, in practice, most of the queue used to execute callbacks is the main queue, for _the existing code_ at least, since that's most likely the current behavior.

Reply:
> Yes, I think that [having a (closure, DispatchQueue) pair object] would be a good idea.
Nice.

> I can definitely start trying it out.
Please don't feel any pressure to do so. It's something I'm fond of and I think could have other useful application, but I'm aware this might not be the best time to go down that rabbit hole.

> I think, in practice, most of the queue used to execute callbacks is the main queue, for the existing code at least, since that's most likely the current behavior.

Interesting, and again we run on the main queue because we use `mainContext` which is associated with the main thread. I think I need to refine my understanding of this mechanism. Where can I look in the codebase to learn more about it?